File: apps/server/internal/market/cache.go
package market

import (
	"context"
	"fmt"
	"log"
	"time"
)

func (s *Service) GetFromCache(ctx context.Context, symbol, exchange string, start, granularity int64) []Candlestick {
	s.cacheMx.RLock()
	cacheKey := fmt.Sprintf("%s-%s-%d-%d", symbol, exchange, granularity, start)
	candles, exists := s.cache[cacheKey]
	s.cacheMx.RUnlock()

	if exists {
		log.Printf("Cache hit: %s\n", cacheKey)
		return candles.Data
	}
	return []Candlestick{}
}

func (s *Service) SaveToCache(ctx context.Context, symbol, exchange string, start, granularity int64, candles []Candlestick) {
	s.cacheMx.Lock()
	cacheKey := fmt.Sprintf("%s-%s-%d-%d", symbol, exchange, granularity, start)
	log.Printf("Cache miss: %s\n", cacheKey)
	s.cache[cacheKey] = CacheCandleBatch{candles, time.Now()}
	s.cacheMx.Unlock()
}

func (s *Service) StartCachePruner(ctx context.Context, ttl time.Duration) {
	log.Println("Starting cache pruner")
	ticker := time.NewTicker(ttl)
	defer ticker.Stop()

	go func() {
		for {
			select {
			case <-ctx.Done():
				return
			case <-ticker.C:
				s.PruneCache(ttl)
			}
		}

	}()
}

func (s *Service) PruneCache(ttl time.Duration) {
	s.cacheMx.Lock()
	defer s.cacheMx.Unlock()

	log.Println("Pruning cache")
	now := time.Now()
	for k, v := range s.cache {
		if now.Sub(v.CreatedAt) > ttl {
			delete(s.cache, k)
		}
	}
}

File: server/src/market/cache.ts
import type { Candlestick } from "./types";

interface CacheCandleBatch {
  data: Candlestick[];
  createdAt: number;
}

// Cache key: <symbol>-<exchange>-<granularity>-<startTime>
export class CandleCache {
  private cache = new Map<string, CacheCandleBatch>();

  constructor(ttlMs: number = 5 * 60 * 1000) {
    const timer = setInterval(() => this.prune(ttlMs), ttlMs);
    timer.unref?.();
  }

  private key(
    symbol: string,
    exchange: string,
    start: number,
    granularity: number,
  ): string {
    return `${symbol}-${exchange}-${granularity}-${start}`;
  }

  get(
    symbol: string,
    exchange: string,
    start: number,
    granularity: number,
  ): Candlestick[] {
    const batch = this.cache.get(this.key(symbol, exchange, start, granularity));
    return batch ? batch.data : [];
  }

  save(
    symbol: string,
    exchange: string,
    start: number,
    granularity: number,
    candles: Candlestick[],
  ): void {
    this.cache.set(this.key(symbol, exchange, start, granularity), {
      data: candles,
      createdAt: Date.now(),
    });
  }

  private prune(ttlMs: number): void {
    const now = Date.now();
    for (const [k, v] of this.cache) {
      if (now - v.createdAt > ttlMs) {
        this.cache.delete(k);
      }
    }
  }
}
